.azure-devops/build/steps/windows/before.yml

```diff
@@ -0,0 +1,75 @@
+      if not exist "${{ parameters.jdkBootDir }}" mkdir ${{ parameters.jdkBootDir }};
+      if not exist "${{ parameters.javaHomeDir }}" mkdir ${{ parameters.javaHomeDir }};
+    displayName: "[Windows Pre] create required directories"
+
+  # download cygwin installer, jdk8u, bootjdk, and visual studio 2013 installer
+  - powershell: |
+      $ProgressPreference = 'SilentlyContinue';
+      if('$(JAVA_TO_BUILD)' -eq 'jdk11u'){
+        $BOOTJDK_VERSION = 10;
+      } elseif('$(JAVA_TO_BUILD)' -eq 'jdk'){
+        $BOOTJDK_VERSION = 14;
+      }
+      Invoke-WebRequest -UseBasicParsing 'https://cygwin.com/setup-x86_64.exe' -OutFile '${{ parameters.dependenciesDir }}\cygwin.exe';
+      Invoke-WebRequest -UseBasicParsing 'https://api.adoptopenjdk.net/v3/binary/latest/8/ga/windows/x64/jdk/hotspot/normal/adoptopenjdk?project=jdk' -OutFile '${{ parameters.dependenciesDir }}\jdk8.zip';
+      if('$(JAVA_TO_BUILD)' -eq 'jdk8u'){
+        Invoke-WebRequest -UseBasicParsing "https://cdn.azul.com/zulu/bin/zulu7.38.0.11-ca-jdk7.0.262-win_x64.zip" -OutFile '${{ parameters.dependenciesDir }}\jdk.zip';
+      } else {
+        Invoke-WebRequest -UseBasicParsing "https://api.adoptopenjdk.net/v3/binary/latest/${BOOTJDK_VERSION}/ga/windows/x64/jdk/hotspot/normal/adoptopenjdk?project=jdk" -OutFile '${{ parameters.dependenciesDir }}\jdk.zip';
+      }
+      Invoke-WebRequest -UseBasicParsing 'https://go.microsoft.com/fwlink/?LinkId=532495' -OutFile '${{ parameters.dependenciesDir }}\vs2013.exe';
+    displayName: "[Windows Pre] download JDKs from AdoptOpenJDK & other dependencies"
+    condition: and(succeeded(), ne(variables.CACHE_RESTORED, 'true'))
+
+  # install cygwin and build dependencies
+  - powershell: |
+      Start-Process -Wait -FilePath '${{ parameters.dependenciesDir }}\cygwin.exe' -ArgumentList '--packages wget,bsdtar,rsync,gnupg,git,autoconf,make,gcc-core,mingw64-x86_64-gcc-core,unzip,zip,cpio,curl,grep,perl --quiet-mode --download --local-install --delete-orphans --site https://mirrors.kernel.org/sourceware/cygwin/ --local-package-dir $(Agent.BuildDirectory)\cygwin_packages --root $(Agent.BuildDirectory)\cygwin64';
+    displayName: "[Windows Pre] install cygwin"
+
+  # add cygwin bin to PATH
+  - script: |
+      echo ##vso[task.setvariable variable=Path]$(Agent.BuildDirectory)\cygwin64\bin;%PATH%
+    displayName: "[Windows Pre] add cygwin to PATH"
+
+  # unzip downloaded JDK zip files to directories, bsdtar is installed by cygwin
+  - bash: |
+      bsdtar --strip-components=1 -xvf $(cygpath -u '${{ parameters.dependenciesDir }}\jdk8.zip') -C '${{ parameters.javaHomeDir }}'
+      bsdtar --strip-components=1 -xvf $(cygpath -u '${{ parameters.dependenciesDir }}\jdk.zip') -C '${{ parameters.jdkBootDir }}'
+    displayName: "[Windows Pre] extracting downloaded JDKs to directories"
+
+  # only install visual studio 2013 when build jdk8 and jdk11
+  - powershell: |
```

Comment:
In the short term future, we'd like to move to vs2017 - but we can tackle that separately

Reply:
that's great the machine comes with vs2017